File: flowcatalyst-typescript/apps/platform/src/domain/principal/principal.ts
/**
 * Principal Aggregate
 *
 * Unified identity model for both users and service accounts.
 * This is the main aggregate for identity management.
 */

import { generate } from '@flowcatalyst/tsid';
import type { PrincipalType } from './principal-type.js';
import type { UserScope } from './user-scope.js';
import type { UserIdentity } from './user-identity.js';
import type { RoleAssignment } from './role-assignment.js';

/**
 * Principal entity - represents a user or service account.
 */
export interface Principal {
	/** TSID primary key */
	readonly id: string;

	/** Type of principal (USER or SERVICE) */
	readonly type: PrincipalType;

	/**
	 * Access scope for user principals.
	 * Determines which clients this user can access.
	 * - ANCHOR: Can access all clients (platform admin)
	 * - PARTNER: Can access explicitly assigned clients
	 * - CLIENT: Can only access their home client
	 *
	 * For SERVICE principals, this may be null.
	 */
	readonly scope: UserScope | null;

	/**
	 * Client this principal belongs to (home client).
	 * - For CLIENT scope: Required, determines their access
	 * - For PARTNER scope: Optional, may have a home client
	 * - For ANCHOR scope: Should be null
	 * - For SERVICE type with client scope: The client the service account belongs to
	 */
	readonly clientId: string | null;

	/**
	 * Application this service account belongs to (for SERVICE type).
	 * When set, this service account was auto-created for the application
	 * and can manage resources prefixed with the application's code.
	 * Null for standalone service accounts or USER type principals.
	 */
	readonly applicationId: string | null;

	/** Display name */
	readonly name: string;

	/** Whether the principal is active */
	readonly active: boolean;

	/** When the principal was created */
	readonly createdAt: Date;

	/** When the principal was last updated */
	readonly updatedAt: Date;

	/** Embedded user identity (for USER type) */
	readonly userIdentity: UserIdentity | null;

	/** Embedded role assignments (denormalized for fast lookup) */
	readonly roles: readonly RoleAssignment[];
}

/**
 * Input for creating a new Principal.
 */
export type NewPrincipal = Omit<Principal, 'createdAt' | 'updatedAt'> & {
	createdAt?: Date;
	updatedAt?: Date;
};

/**
 * Create a new user principal.
 */
export function createUserPrincipal(params: {
	name: string;
	scope: UserScope;
	clientId: string | null;
	userIdentity: UserIdentity;
}): NewPrincipal {
	return {
		id: generate('PRINCIPAL'),
		type: 'USER',
		scope: params.scope,
		clientId: params.clientId,
		applicationId: null,
		name: params.name,
		active: true,
		userIdentity: params.userIdentity,
		roles: [],
	};
}

/**
 * Get role names from a principal as a Set.
 */
export function getRoleNames(principal: Principal): ReadonlySet<string> {
	return new Set(principal.roles.map((r) => r.roleName));
}

/**
 * Check if a principal has a specific role.
 */
export function hasRole(principal: Principal, roleName: string): boolean {
	return principal.roles.some((r) => r.roleName === roleName);
}

/**
 * Update a principal with new values.
 */
export function updatePrincipal(
	principal: Principal,
	updates: Partial<Pick<Principal, 'name' | 'active' | 'scope' | 'clientId' | 'userIdentity' | 'roles'>>,
): Principal {
	return {
		...principal,
		...updates,
		updatedAt: new Date(),
	};
}

/**
 * Assign roles to a principal, replacing existing roles.
 */
export function assignRoles(principal: Principal, roleAssignments: readonly RoleAssignment[]): Principal {
	return {
		...principal,
		roles: roleAssignments,
		updatedAt: new Date(),
	};
}

File: core/flowcatalyst-platform/src/main/java/tech/flowcatalyst/platform/principal/Principal.java
package tech.flowcatalyst.platform.principal;

import java.time.Instant;
import java.util.ArrayList;
import java.util.List;
import java.util.Set;
import java.util.stream.Collectors;

/**
 * Unified identity model for both users and service accounts.
 * Follows the architecture documented in docs/auth-architecture.md
 */
public class Principal {

    public String id;

    public PrincipalType type;

    /**
     * Access scope for user principals.
     * Determines which clients this user can access.
     * - ANCHOR: Can access all clients (platform admin)
     * - PARTNER: Can access explicitly assigned clients
     * - CLIENT: Can only access their home client
     *
     * For SERVICE principals, this is typically null.
     */
    public UserScope scope;

    /**
     * Client this principal belongs to (home client).
     * - For CLIENT scope: Required, determines their access
     * - For PARTNER scope: Optional, may have a home client
     * - For ANCHOR scope: Should be null
     * - For SERVICE type with client scope: The client the service account belongs to
     */
    public String clientId;


    /**
     * Scope for application management access.
     * Determines which applications this principal can manage.
     * - ALL: Can manage all applications (platform admins)
     * - SPECIFIC: Can only manage applications in managedApplicationIds
     * - NONE: Cannot manage any applications (default)
     */
    public ManagedApplicationScope managedApplicationScope = ManagedApplicationScope.NONE;

    /**
     * Applications this principal can manage (create roles, permissions, event types, etc.).
     * Interpretation depends on managedApplicationScope:
     * - ALL: This list is ignored, can manage all applications
     * - SPECIFIC: Can only manage applications in this list
     * - NONE: Cannot manage any applications (list should be empty)
     */
    public List<String> managedApplicationIds = new ArrayList<>();

    public String name;

    public boolean active = true;

    public Instant createdAt = Instant.now();

    public Instant updatedAt = Instant.now();

    /**
     * Embedded user identity (for USER type).
     * Fields are stored as flat columns in the database but assembled here for convenience.
     */
    public UserIdentity userIdentity;

    /**
     * Foreign key to the ServiceAccount entity (for SERVICE type).
     * This is the primary way to link a Principal to a ServiceAccount.
     * The ServiceAccount entity contains webhook credentials and other metadata.
     */
    public String serviceAccountId;


    /**
     * Embedded role assignments (denormalized for MongoDB).
     * This is the source of truth for principal roles.
     */
    public List<RoleAssignment> roles = new ArrayList<>();

    public Principal() {
    }

    /**
     * Get role names as a set for quick lookup.
     */
    public Set<String> getRoleNames() {
        return roles.stream()
            .map(r -> r.roleName)
            .filter(java.util.Objects::nonNull)
            .collect(Collectors.toSet());
    }

    /**
     * Check if principal has a specific role.
     */
    public boolean hasRole(String roleName) {
        return roles.stream().anyMatch(r -> r.roleName.equals(roleName));
    }

    /**
     * Embedded role assignment.
     */
    public static class RoleAssignment {
        public String roleName;
        public String assignmentSource;
        public Instant assignedAt;

        public RoleAssignment() {
        }

        public RoleAssignment(String roleName, String assignmentSource) {
            this.roleName = roleName;
            this.assignmentSource = assignmentSource;
            this.assignedAt = Instant.now();
        }

        public RoleAssignment(String roleName, String assignmentSource, Instant assignedAt) {
            this.roleName = roleName;
            this.assignmentSource = assignmentSource;
            this.assignedAt = assignedAt;
        }
    }
}
